runtime/j9vm/valhallavmi.cpp

```diff
@@ -91,24 +110,33 @@ JVM_NewNullRestrictedArray(JNIEnv *env, jclass componentType, jint length)
 		goto done;
 	}
 
-	if (!(J9_IS_J9CLASS_VALUETYPE(ramClass) && J9_IS_J9CLASS_ALLOW_DEFAULT_VALUE(ramClass))) {
+	if (!J9_IS_J9CLASS_VALUETYPE(ramClass)) {
 		vmFuncs->setCurrentException(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, NULL);
 		goto done;
 	}
 
-	if (NULL == J9CLASS_GET_NULLRESTRICTED_ARRAY(ramClass)) {
+	if (isNullRestricted && !J9_IS_J9CLASS_ALLOW_DEFAULT_VALUE(ramClass)) {
+		vmFuncs->setCurrentException(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, NULL);
+		goto done;
+	}
+
+	if (NULL == (isNullRestricted ? J9CLASS_GET_NULLRESTRICTED_ARRAY(ramClass) : ramClass->arrayClass)) {
+		UDATA options = isNullRestricted ? J9_FINDCLASS_FLAG_CLASS_OPTION_NULL_RESTRICTED_ARRAY : 0;
 		J9ROMArrayClass *arrayOfObjectsROMClass = (J9ROMArrayClass *)J9ROMIMAGEHEADER_FIRSTCLASS(vm->arrayROMClasses);
-		vmFuncs->internalCreateArrayClassWithOptions(
-			currentThread, arrayOfObjectsROMClass, ramClass, J9_FINDCLASS_FLAG_CLASS_OPTION_NULL_RESTRICTED_ARRAY);
+		vmFuncs->internalCreateArrayClassWithOptions(currentThread, arrayOfObjectsROMClass, ramClass, options);
 		if (NULL != currentThread->currentException) {
 			goto done;
 		}
 		ramClass = VM_VMHelpers::currentClass(ramClass);
 	}
 
-	newArray = vm->memoryManagerFunctions->J9AllocateIndexableObject(
-			currentThread, J9CLASS_GET_NULLRESTRICTED_ARRAY(ramClass), length, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
-
+	if (isNullRestricted) {
```

Comment:
can we save the return value of `internalCreateArrayClassWithOptions` so we already have the right array in hand and avoid the `if (isNullRestricted) {` test here?